File: substrate/collective_floating_cohesive_prompt.py
"""Collective floating deep research → cohesive unit prompt (pure).

Multi-select floating/sub-agent instances and prompt as one cohesive unit.
live_dispatched is always False. Context cards are caller-supplied only.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

MemberStatus = Literal["proposed", "open", "completed", "closed"]


class CollectiveFloatingCohesivePromptError(ValueError):
    """Fail-closed validation for cohesive floating prompt packs."""


@dataclass(frozen=True)
class CohesiveUnitPromptIntent:
    parent_asset_id: str
    instance_ids: tuple[str, ...]
    cohesive_prompt: str
    context_cards: tuple[str, ...]
    member_count: int
    operator_ack: bool
    pack_ready: bool
    live_dispatched: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "parent_asset_id": self.parent_asset_id,
            "instance_ids": list(self.instance_ids),
            "cohesive_prompt": self.cohesive_prompt,
            "context_cards": list(self.context_cards),
            "member_count": self.member_count,
            "operator_ack": self.operator_ack,
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "notes": list(self.notes),
            "authority": "collective_floating_cohesive_prompt_advisory",
        }


def _require_bool(value: object, *, field: str) -> bool:
    if not isinstance(value, bool):
        raise CollectiveFloatingCohesivePromptError(
            f"{field} must be an explicit boolean"
        )
    return value


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise CollectiveFloatingCohesivePromptError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def build_collective_floating_cohesive_prompt(
    members: object,
    *,
    cohesive_prompt: object,
    operator_ack: object,
    extra_context: object | None = None,
) -> CohesiveUnitPromptIntent:
    """Build cohesive-unit prompt pack. Never live-dispatches."""
    ack = _require_bool(operator_ack, field="operator_ack")
    prompt = _require_nonempty(cohesive_prompt, field="cohesive_prompt")
    if not isinstance(members, list) or len(members) < 2:
        raise CollectiveFloatingCohesivePromptError(
            "cohesive unit requires at least 2 members"
        )

    first = members[0]
    parent = _require_nonempty(
        first.get("parent_asset_id") if isinstance(first, dict) else None,
        field="members[0].parent_asset_id",
    )
    ids: list[str] = []
    context_cards: list[str] = []
    notes: list[str] = [
        "live_dispatched=false — cohesive pack intent only",
        "context cards are caller-supplied only (no invent)",
    ]

    for i, m in enumerate(members):
        if not isinstance(m, dict):
            raise CollectiveFloatingCohesivePromptError(
                f"members[{i}] must be an object"
            )
        iid = _require_nonempty(
            m.get("instance_id"), field=f"members[{i}].instance_id"
        )
        p = _require_nonempty(
            m.get("parent_asset_id"), field=f"members[{i}].parent_asset_id"
        )
        if p != parent:
            raise CollectiveFloatingCohesivePromptError(
                "cohesive unit requires same parent_asset_id"
            )
        status = m.get("status")
        if status not in ("proposed", "open", "completed"):
            raise CollectiveFloatingCohesivePromptError(
                f"members[{i}] status must be proposed|open|completed (not closed)"
            )
        ids.append(iid)

        highlight = m.get("highlight")
        if highlight is not None:
            if not isinstance(highlight, str) or not highlight.strip():
                raise CollectiveFloatingCohesivePromptError(
                    f"members[{i}].highlight must be non-empty string when set"
                )
            context_cards.append(f"[{iid} highlight] {highlight.strip()}")

        prior = m.get("prior_prompt")
        if prior is not None:
            if not isinstance(prior, str) or not prior.strip():
                raise CollectiveFloatingCohesivePromptError(
                    f"members[{i}].prior_prompt must be non-empty string when set"
                )
            context_cards.append(f"[{iid} prior_prompt] {prior.strip()}")

        ctx = m.get("context")
        if ctx is not None:
            if not isinstance(ctx, list):
                raise CollectiveFloatingCohesivePromptError(
                    f"members[{i}].context must be string[] when set"
                )
            for j, c in enumerate(ctx):
                if not isinstance(c, str) or not c.strip():
                    raise CollectiveFloatingCohesivePromptError(
                        f"members[{i}].context[{j}] must be non-empty string"
                    )
                context_cards.append(f"[{iid}] {c.strip()}")

    seen: set[str] = set()
    unique: list[str] = []
    for iid in ids:
        if iid in seen:
            continue
        seen.add(iid)
        unique.append(iid)
    if len(unique) < 2:
        raise CollectiveFloatingCohesivePromptError(
            "cohesive unit requires at least 2 distinct instance_ids"
        )

    if extra_context is not None:
        if not isinstance(extra_context, list):
            raise CollectiveFloatingCohesivePromptError(
                "extra_context must be string[] or null"
            )
        for j, c in enumerate(extra_context):
            if not isinstance(c, str) or not c.strip():
                raise CollectiveFloatingCohesivePromptError(
                    f"extra_context[{j}] must be non-empty string"
                )
            context_cards.append(c.strip())

    if not context_cards:
        notes.append(
            "no context cards supplied — prompt pack scaffold only (no invent content)"
        )
    else:
        notes.append(f"context_cards={len(context_cards)} caller-supplied only")

    pack_ready = ack is True
    if not pack_ready:
        notes.append(
            "pack_ready=false — operator_ack required before dispatch gate"
        )
    else:
        notes.append(
            "pack_ready=true — still live_dispatched=false (pure layer never dispatches)"
        )
    notes.append("live_dispatched=false")

    return CohesiveUnitPromptIntent(
        parent_asset_id=parent,
        instance_ids=tuple(unique),
        cohesive_prompt=prompt,
        context_cards=tuple(context_cards),
        member_count=len(unique),
        operator_ack=ack,
        pack_ready=pack_ready,
        live_dispatched=False,
        notes=tuple(notes),
        authority="collective_floating_cohesive_prompt_advisory",
    )


__all__ = [
    "CohesiveUnitPromptIntent",
    "CollectiveFloatingCohesivePromptError",
    "build_collective_floating_cohesive_prompt",
]

File: apps/reading/src/api/collectiveFloatingCohesivePrompt.ts
/**
 * Collective floating deep research → cohesive unit prompt (pure).
 *
 * Operator vision: multi-select floating/sub-agent deep research instances
 * and prompt them as one cohesive unit (not only post-hoc analysis merge).
 *
 * live_dispatched is always false in this pure layer — pack intent only.
 * Never invents findings or context from instances; caller-supplied only.
 */

export type CohesiveMemberStatus =
  | "proposed"
  | "open"
  | "completed"
  | "closed";

export interface CohesiveFloatingMember {
  instance_id: string;
  parent_asset_id: string;
  status: CohesiveMemberStatus;
  highlight?: string;
  prior_prompt?: string;
  /** Optional operator-supplied context cards — never invented. */
  context?: string[];
}

export interface CohesiveUnitPromptIntent {
  parent_asset_id: string;
  instance_ids: string[];
  cohesive_prompt: string;
  /** Flattened caller-supplied context only. */
  context_cards: string[];
  member_count: number;
  operator_ack: boolean;
  /**
   * True only when operator_ack and ≥2 distinct members share parent.
   * Still does not authorize live dispatch.
   */
  pack_ready: boolean;
  /** Always false — pure intent never dispatches multi-agent pack. */
  live_dispatched: false;
  notes: string[];
  authority: "collective_floating_cohesive_prompt_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Build a cohesive-unit prompt pack from ≥2 floating research instances.
 * Never live-dispatches; never invents context.
 */
export function buildCollectiveFloatingCohesivePrompt(
  members: CohesiveFloatingMember[],
  input: {
    cohesive_prompt: string;
    operator_ack: boolean;
    /** Optional extra context cards from the operator. */
    extra_context?: string[] | null;
  },
): CohesiveUnitPromptIntent {
  if (!Array.isArray(members) || members.length < 2) {
    throw new Error("cohesive unit requires at least 2 members");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const cohesive_prompt = requireNonEmpty(
    input.cohesive_prompt,
    "cohesive_prompt",
  );

  const parent = requireNonEmpty(
    members[0]?.parent_asset_id,
    "members[0].parent_asset_id",
  );
  const ids: string[] = [];
  const context_cards: string[] = [];
  const notes: string[] = [
    "live_dispatched=false — cohesive pack intent only",
    "context cards are caller-supplied only (no invent)",
  ];

  for (let i = 0; i < members.length; i++) {
    const m = members[i];
    if (!m || typeof m !== "object") {
      throw new Error(`members[${i}] must be an object`);
    }
    const id = requireNonEmpty(m.instance_id, `members[${i}].instance_id`);
    const p = requireNonEmpty(
      m.parent_asset_id,
      `members[${i}].parent_asset_id`,
    );
    if (p !== parent) {
      throw new Error("cohesive unit requires same parent_asset_id");
    }
    if (
      m.status !== "proposed" &&
      m.status !== "open" &&
      m.status !== "completed"
    ) {
      throw new Error(
        `members[${i}] status must be proposed|open|completed (not closed)`,
      );
    }
    ids.push(id);

    if (m.highlight != null) {
      if (typeof m.highlight !== "string" || !m.highlight.trim()) {
        throw new Error(
          `members[${i}].highlight must be non-empty string when set`,
        );
      }
      context_cards.push(`[${id} highlight] ${m.highlight.trim()}`);
    }
    if (m.prior_prompt != null) {
      if (typeof m.prior_prompt !== "string" || !m.prior_prompt.trim()) {
        throw new Error(
          `members[${i}].prior_prompt must be non-empty string when set`,
        );
      }
      context_cards.push(`[${id} prior_prompt] ${m.prior_prompt.trim()}`);
    }
    if (m.context != null) {
      if (!Array.isArray(m.context)) {
        throw new Error(`members[${i}].context must be string[] when set`);
      }
      for (let j = 0; j < m.context.length; j++) {
        const c = m.context[j];
        if (typeof c !== "string" || !c.trim()) {
          throw new Error(
            `members[${i}].context[${j}] must be non-empty string`,
          );
        }
        context_cards.push(`[${id}] ${c.trim()}`);
      }
    }
  }

  const seen = new Set<string>();
  const unique = ids.filter((id) => {
    if (seen.has(id)) return false;
    seen.add(id);
    return true;
  });
  if (unique.length < 2) {
    throw new Error("cohesive unit requires at least 2 distinct instance_ids");
  }

  if (input.extra_context != null) {
    if (!Array.isArray(input.extra_context)) {
      throw new Error("extra_context must be string[] or null");
    }
    for (let j = 0; j < input.extra_context.length; j++) {
      const c = input.extra_context[j];
      if (typeof c !== "string" || !c.trim()) {
        throw new Error(`extra_context[${j}] must be non-empty string`);
      }
      context_cards.push(c.trim());
    }
  }

  if (context_cards.length === 0) {
    notes.push(
      "no context cards supplied — prompt pack scaffold only (no invent content)",
    );
  } else {
    notes.push(`context_cards=${context_cards.length} caller-supplied only`);
  }

  const pack_ready = input.operator_ack === true;
  if (!pack_ready) {
    notes.push("pack_ready=false — operator_ack required before dispatch gate");
  } else {
    notes.push(
      "pack_ready=true — still live_dispatched=false (pure layer never dispatches)",
    );
  }
  notes.push("live_dispatched=false");

  return {
    parent_asset_id: parent,
    instance_ids: unique,
    cohesive_prompt,
    context_cards,
    member_count: unique.length,
    operator_ack: input.operator_ack,
    pack_ready,
    live_dispatched: false,
    notes,
    authority: "collective_floating_cohesive_prompt_advisory",
  };
}

export function formatCohesivePromptSummary(
  intent: CohesiveUnitPromptIntent,
): string {
  return (
    `cohesive pack · members=${intent.member_count} · ` +
    `context=${intent.context_cards.length} · pack_ready=${intent.pack_ready} · ` +
    `live_dispatched=false`
  );
}
